### PR TITLE
[Merged by Bors] - feat: port algebra.hom.centroid

### PR DESCRIPTION
**Pull request by maxwell-thum**

### CODE REVIEW

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+* [Jacobson, Structure of Rings][Jacobson1956]
+* [McCrimmon, A taste of Jordan algebras][mccrimmon2004]
+
+## Tags
+
+centroid
+-/
+
+-- Porting note: This is so that `instance : Ring (CentroidHom α)` doesn't time out.
+set_option maxHeartbeats 1000000
```
Comment (jcommelin) [inline Mathlib/Algebra/Hom/Centroid.lean:48]: Please move it to that particular instance using
```
set_option ... in
```
syntax

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+@[simp]
+theorem id_apply (a : α) : CentroidHom.id α a = a :=
+  rfl
+#align centroid_hom.id_apply CentroidHom.id_apply
+
+/-- Composition of `centroid_hom`s as a `centroid_hom`. -/
+def comp (g f : CentroidHom α) : CentroidHom α :=
+  {
+    g.toAddMonoidHom.comp
+      f.toAddMonoidHom with
```
Comment (jcommelin) [inline Mathlib/Algebra/Hom/Centroid.lean:197]:
```suggestion
  { g.toAddMonoidHom.comp f.toAddMonoidHom with
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+      map_mul_right' := fun _a _b ↦ (zero_mul _).symm }⟩
+
+instance : One (CentroidHom α) :=
+  ⟨CentroidHom.id α⟩
+
+instance : Add (CentroidHom α) :=
+  ⟨fun f g ↦
+    {
+      (f + g : α →+
+          α) with
```
Comment (jcommelin) [inline Mathlib/Algebra/Hom/Centroid.lean:255]:
```suggestion
    { (f + g : α →+ α) with
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+        show f (a * b) + g (a * b) = (f a + g a) * b
+        simp [map_mul_right, add_mul] }⟩
+
+instance : Mul (CentroidHom α) :=
+  ⟨comp⟩
+
+instance hasNsmul : SMul ℕ (CentroidHom α) :=
+  ⟨fun n f ↦
+    {
+      ((SMul.smul n f) : α →+ α) with
```
Comment (jcommelin) [inline Mathlib/Algebra/Hom/Centroid.lean:269]:
```suggestion
    { ((SMul.smul n f) : α →+ α) with
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+          rw [map_mul_left f, ← mul_smul_comm]
+        map_mul_right' := fun a b ↦ by
+          change n • f (a * b) = n • f a * b
+          rw [map_mul_right f, ← smul_mul_assoc] }⟩
+#align centroid_hom.has_nsmul CentroidHom.hasNsmul
+
+instance hasNpowNat : Pow (CentroidHom α) ℕ :=
+  ⟨fun f n ↦
+    {
+      (f.toEnd ^ n :
```
Comment (jcommelin) [inline Mathlib/Algebra/Hom/Centroid.lean:281]:
```suggestion
    { (f.toEnd ^ n :
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+        simp [map_mul_left]
+      map_mul_right' := fun a b ↦ by
+        change -f (a * b) = (-f a) * b
+        simp [map_mul_right] }⟩
+
+instance : Sub (CentroidHom α) :=
+  ⟨fun f g ↦
+    {
+      (f - g : α →+
+          α) with
```
Comment (jcommelin) [inline Mathlib/Algebra/Hom/Centroid.lean:431]:
```suggestion
    { (f - g : α →+ α) with
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+@[simp]
+theorem to_End_add (x y : CentroidHom α) : (x + y).toEnd = x.toEnd + y.toEnd :=
+  rfl
+#align centroid_hom.to_End_add CentroidHom.to_End_add
+
+theorem to_End_nsmul (x : CentroidHom α) (n : ℕ) : (n • x).toEnd = n • x.toEnd :=
+  rfl
+#align centroid_hom.to_End_nsmul CentroidHom.to_End_nsmul
+
+-- Porting note: I guess the porter has naming issues still
```
Comment (jcommelin) [inline Mathlib/Algebra/Hom/Centroid.lean:364]: What does this mean?

Comment (maxwell-thum) [reply Mathlib/Algebra/Hom/Centroid.lean:358]: I phrased it poorly, but I was referring to the fact that I had to change `coe_to_add_monoid_hom_injective.AddCommMonoid` to `coe_to_add_monoid_hom_injective.addCommMonoid`

File: Mathlib/Algebra/Hom/Centroid.lean (hunk 1 above)
Comment (maxwell-thum) [reply Mathlib/Algebra/Hom/Centroid.lean]: I'm not exactly sure what you mean; is the most recent commit what you're looking for?

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+# Centroid homomorphisms
+
+Let `A` be a (non unital, non associative) algebra. The centroid of `A` is the set of linear maps
+`T` on `A` such that `T` commutes with left and right multiplication, that is to say, for all `a`
+and `b` in `A`,
+$$
+T(ab) = (Ta)b, T(ab) = a(Tb).
+$$
+In mathlib we call elements of the centroid "centroid homomorphisms" (`centroid_hom`) in keeping
+with `add_monoid_hom` etc.
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:24]:
```suggestion
In mathlib we call elements of the centroid "centroid homomorphisms" (`CentroidHom`) in keeping
with `AddMonoidHom` etc.
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+
+We use the `fun_like` design, so each type of morphisms has a companion typeclass which is meant to
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:26]:
```suggestion
We use the `FunLike` design, so each type of morphisms has a companion typeclass which is meant to
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 10]
```diff
@@ -0,0 +1,530 @@
...
+be satisfied by itself and all stricter types.
+
+## Types of morphisms
+
+* `centroid_hom`: Maps which preserve left and right multiplication.
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:31]:
```suggestion
* `CentroidHom`: Maps which preserve left and right multiplication.
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 11]
```diff
@@ -0,0 +1,530 @@
...
+
+## Typeclasses
+
+* `centroid_hom_class`
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:35]:
```suggestion
* `CentroidHomClass`
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 12]
```diff
@@ -0,0 +1,530 @@
...
+  rfl
+#align centroid_hom.coe_to_add_monoid_hom CentroidHom.coe_to_add_monoid_hom
+
+@[simp]
+theorem to_add_monoid_hom_eq_coe (f : CentroidHom α) : f.toAddMonoidHom = f :=
+  rfl
+#align centroid_hom.to_add_monoid_hom_eq_coe CentroidHom.to_add_monoid_hom_eq_coe
+
+theorem coe_to_add_monoid_hom_injective : Injective ((↑) : CentroidHom α → α →+ α) := fun _f _g h =>
+  ext fun a ↦
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:135]:
```suggestion
theorem coe_to_add_monoid_hom_injective : Injective ((↑) : CentroidHom α → α →+ α) := 
  fun _f _g h => ext fun a ↦
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+#align centroid_hom.coe_to_add_monoid_hom_injective CentroidHom.coe_to_add_monoid_hom_injective
+
+/-- Turn a centroid homomorphism into an additive monoid endomorphism. -/
+def toEnd (f : CentroidHom α) : AddMonoid.End α :=
+  (f : α →+ α)
+#align centroid_hom.to_End CentroidHom.toEnd
+
+theorem to_End_injective : Injective (CentroidHom.toEnd : CentroidHom α → AddMonoid.End α) :=
+  coe_to_add_monoid_hom_injective
+#align centroid_hom.to_End_injective CentroidHom.to_End_injective
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:147]:
```suggestion
theorem toEnd_injective : Injective (CentroidHom.toEnd : CentroidHom α → AddMonoid.End α) :=
  coe_to_add_monoid_hom_injective
#align centroid_hom.to_End_injective CentroidHom.toEnd_injective
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 14]
```diff
@@ -0,0 +1,530 @@
...
+
+@[simp, norm_cast]
+theorem coe_id : ⇑(CentroidHom.id α) = id :=
+  rfl
+#align centroid_hom.coe_id CentroidHom.coe_id
+
+@[simp, norm_cast]
+theorem coe_to_add_monoid_hom_id : (CentroidHom.id α : α →+ α) = AddMonoidHom.id α :=
+  rfl
+#align centroid_hom.coe_to_add_monoid_hom_id CentroidHom.coe_to_add_monoid_hom_id
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:187]:
```suggestion
theorem toAddMonoidHom_id : (CentroidHom.id α : α →+ α) = AddMonoidHom.id α :=
  rfl
#align centroid_hom.coe_to_add_monoid_hom_id CentroidHom.toAddMonoidHom_id
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 15]
```diff
@@ -0,0 +1,530 @@
...
+
+@[simp]
+theorem comp_apply (g f : CentroidHom α) (a : α) : g.comp f a = g (f a) :=
+  rfl
+#align centroid_hom.comp_apply CentroidHom.comp_apply
+
+@[simp, norm_cast]
+theorem coe_comp_add_monoid_hom (g f : CentroidHom α) : (g.comp f : α →+ α) = (g : α →+ α).comp f :=
+  rfl
+#align centroid_hom.coe_comp_add_monoid_hom CentroidHom.coe_comp_add_monoid_hom
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:217]:
```suggestion
theorem coe_comp_addMonoidHom (g f : CentroidHom α) : (g.comp f : α →+ α) = (g : α →+ α).comp f :=
  rfl
#align centroid_hom.coe_comp_add_monoid_hom CentroidHom.coe_comp_add_monoid_hom
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 16]
```diff
@@ -0,0 +1,530 @@
...
+#align centroid_hom.to_End_zero CentroidHom.to_End_zero
+
+@[simp]
+theorem to_End_add (x y : CentroidHom α) : (x + y).toEnd = x.toEnd + y.toEnd :=
+  rfl
+#align centroid_hom.to_End_add CentroidHom.to_End_add
+
+theorem to_End_nsmul (x : CentroidHom α) (n : ℕ) : (n • x).toEnd = n • x.toEnd :=
+  rfl
+#align centroid_hom.to_End_nsmul CentroidHom.to_End_nsmul
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:359]:
```suggestion
@[simp]
theorem toEnd_zero : (0 : CentroidHom α).toEnd = 0 :=
  rfl
#align centroid_hom.to_End_zero CentroidHom.toEnd_zero

@[simp]
theorem toEnd_add (x y : CentroidHom α) : (x + y).toEnd = x.toEnd + y.toEnd :=
  rfl
#align centroid_hom.to_End_add CentroidHom.toEnd_add

theorem toEnd_nsmul (x : CentroidHom α) (n : ℕ) : (n • x).toEnd = n • x.toEnd :=
  rfl
#align centroid_hom.to_End_nsmul CentroidHom.toEnd_nsmul
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 17]
```diff
@@ -0,0 +1,530 @@
...
+
+@[simp]
+theorem to_End_pow (x : CentroidHom α) (n : ℕ) : (x ^ n).toEnd = x.toEnd ^ n :=
+  rfl
+#align centroid_hom.to_End_pow CentroidHom.to_End_pow
+
+@[simp, norm_cast]
+theorem to_End_nat_cast (n : ℕ) : (n : CentroidHom α).toEnd = ↑n :=
+  rfl
+#align centroid_hom.to_End_nat_cast CentroidHom.to_End_nat_cast
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:396]:
```suggestion
@[simp]
theorem toEnd_one : (1 : CentroidHom α).toEnd = 1 :=
  rfl
#align centroid_hom.to_End_one CentroidHom.toEnd_one

@[simp]
theorem toEnd_mul (x y : CentroidHom α) : (x * y).toEnd = x.toEnd * y.toEnd :=
  rfl
#align centroid_hom.to_End_mul CentroidHom.toEnd_mul

@[simp]
theorem toEnd_pow (x : CentroidHom α) (n : ℕ) : (x ^ n).toEnd = x.toEnd ^ n :=
  rfl
#align centroid_hom.to_End_pow CentroidHom.toEnd_pow

@[simp, norm_cast]
theorem toEnd_nat_cast (n : ℕ) : (n : CentroidHom α).toEnd = ↑n :=
  rfl
#align centroid_hom.to_End_nat_cast CentroidHom.toEnd_nat_cast
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 18]
```diff
@@ -0,0 +1,530 @@
...
+#align centroid_hom.to_End_neg CentroidHom.to_End_neg
+
+@[simp]
+theorem to_End_sub (x y : CentroidHom α) : (x - y).toEnd = x.toEnd - y.toEnd :=
+  rfl
+#align centroid_hom.to_End_sub CentroidHom.to_End_sub
+
+theorem to_End_zsmul (x : CentroidHom α) (n : ℤ) : (n • x).toEnd = n • x.toEnd :=
+  rfl
+#align centroid_hom.to_End_zsmul CentroidHom.to_End_zsmul
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:473]:
```suggestion
@[simp]
theorem toEnd_neg (x : CentroidHom α) : (-x).toEnd = -x.toEnd :=
  rfl
#align centroid_hom.to_End_neg CentroidHom.toEnd_neg

@[simp]
theorem toEnd_sub (x y : CentroidHom α) : (x - y).toEnd = x.toEnd - y.toEnd :=
  rfl
#align centroid_hom.to_End_sub CentroidHom.toEnd_sub

theorem toEnd_zsmul (x : CentroidHom α) (n : ℤ) : (n • x).toEnd = n • x.toEnd :=
  rfl
#align centroid_hom.to_End_zsmul CentroidHom.toEnd_zsmul
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 19]
```diff
@@ -0,0 +1,530 @@
...
+
+@[simp]
+theorem sub_apply (f g : CentroidHom α) (a : α) : (f - g) a = f a - g a :=
+  rfl
+#align centroid_hom.sub_apply CentroidHom.sub_apply
+
+@[simp, norm_cast]
+theorem to_End_int_cast (z : ℤ) : (z : CentroidHom α).toEnd = ↑z :=
+  rfl
+#align centroid_hom.to_End_int_cast CentroidHom.to_End_int_cast
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:502]:
```suggestion
theorem toEnd_int_cast (z : ℤ) : (z : CentroidHom α).toEnd = ↑z :=
  rfl
#align centroid_hom.to_End_int_cast CentroidHom.toEnd_int_cast
```

File: Mathlib/Algebra/Hom/Centroid.lean [hunk 20]
```diff
@@ -0,0 +1,530 @@
...
+
+set_option maxHeartbeats 1000000
```
Comment (ChrisHughes24) [inline Mathlib/Algebra/Hom/Centroid.lean:504]:
```suggestion
set_option maxHeartbeats 1000000 in
```

**Pull request closed without merging**